lib/improver/tests/metadata/test_check_datatypes.py

```diff
@@ -0,0 +1,307 @@
+        self.cube.coord("time").points = (
+             self.cube.coord("time").points.astype(np.float64))
+        enforce.check_cube_not_float64(self.cube)
+
+    def test_float64_cube_forecast_ref_time_coord_points_ok(self):
+        """Test a pass of a cube with 64 bit fcast ref time coord points."""
+        frt_coord = iris.coords.AuxCoord(
+            [np.float64(min(self.cube.coord("time").points))],
+            standard_name="forecast_reference_time")
+        self.cube.add_aux_coord(frt_coord)
+        enforce.check_cube_not_float64(self.cube)
+
+
+class Test_check_time_coordinate_metadata(IrisTest):
+    """Test check_time_coordinate_metatadata function"""
+
+    def test_basic(self):
+        """Test success"""
+        cube = set_up_variable_cube(np.ones((4, 4), dtype=np.float32))
+        enforce.check_time_coordinate_metadata(cube)
+
+    def test_fails_wrong_datatype(self):
+        """Test failure if any coordinate datatype is wrong"""
+        cube = set_up_variable_cube(np.ones((4, 4), dtype=np.float32))
+        cube.coord("time").points = (
+            cube.coord("time").points.astype(np.float64))
+        msg = 'Coordinate time does not match required standard'
+        with self.assertRaisesRegex(ValueError, msg):
+            enforce.check_time_coordinate_metadata(cube)
+
+    def test_fails_wrong_units(self):
+        """Test failure if any coordinate unit is wrong"""
+        cube = set_up_variable_cube(278*np.ones((4, 4), dtype=np.float32))
+        cube.coord("forecast_period").convert_units("hours")
+        msg = 'Coordinate forecast_period does not match required standard'
+        with self.assertRaisesRegex(ValueError, msg):
+            enforce.check_time_coordinate_metadata(cube)
+
+
+class Test__construct_object_list(IrisTest):
+    """Test the private _construct_object_list method"""
+
+    def setUp(self):
+        """Make a template cube"""
+        self.cube = set_up_variable_cube(
+            278*np.ones((3, 4, 4), dtype=np.float32))
+
+    def test_basic(self):
+        """Test it works for all coordinates"""
+        expected_result = {
+            self.cube, self.cube.coord('realization'),
+            self.cube.coord('latitude'), self.cube.coord('longitude'),
+            self.cube.coord('time'), self.cube.coord('forecast_period'),
+            self.cube.coord('forecast_reference_time')}
+        result = enforce._construct_object_list(self.cube, None)
+        self.assertSetEqual(set(result), expected_result)
+
+    def test_subset(self):
+        """Test it works on a subset and ignores any missing coordinates"""
+        expected_result = {
+            self.cube, self.cube.coord('realization'), self.cube.coord('time')}
+        result = enforce._construct_object_list(
+            self.cube, ['realization', 'time', 'kittens'])
+        self.assertSetEqual(set(result), expected_result)
+
+
+class Test_check_datatypes(IrisTest):
```

Comment:
To be through we could also check a datatype taken from TIME_METADATA and another one from INTEGER_QUANTITIES, just to check we are accessing these successfully.

Reply:
There are checks for `forecast_period` datatype within this class.

In general we should be checking results not methods.  @TomekTrzeciak 's comments make it clear a dictionary implementation is not ideal, and I have yet to go through the detail but it may be possible to get rid of the dictionaries completely in this PR.  So checking a representative set of diagnostics is good, but we don't need tests that assume that specific implementation.

Reply:
Now got rid of dictionaries.